doc: Getting started guide (#51)
allen-cell-animated/colorizer-data: documentation/GETTING_STARTED.md

> # Getting Started: how to process data for Timelapse Feature Explorer

Author: Markdown preview here: https://github.com/allen-cell-animated/colorizer-data/blob/doc/getting-started-guide/documentation/GETTING_STARTED.md

Author: Also note that some of the links are broken since they point to where date *will be* once this is merged. I'll fix this when I have time so they work for now and then revert the change before it goes to main.

The [Timelapse Feature Explorer (TFE)](https://timelapse.allencell.org) is a web-based application designed for the interactive visualization and analysis of segmented time-series microscopy data! Data needs to be processed into a specific format to be loaded into the viewer.

In this tutorial, you'll learn how to prepare your data for the Timelapse Feature Explorer.

## 1. Terms

A few key terms:

> - **Dataset**: A dataset is a single time-series, and can have any number of tracked objects and features.

Reviewer: I don't know what size of data people might bring, but it's actually limited to the 32-bit pixel size of a png for our indexing.
```suggestion
- **Dataset**: A dataset is a single time-series, and can have any number of tracked objects and features. (up to about 16.7 million unique tracked objects)
```

- **Raw dataset**: The raw data that you have collected or generated, before processing into the TFE format.
- **Collection**: An arbitrary grouping of datasets.
- **Object ID**: An ID associated with a single segmentation at a single timepoint. In the TFE-accepted format, object IDs must be sequential, starting from 0, and be unique across the whole dataset.

> - **Track ID**: An identifier for a unique set of objects, linking their object IDs across timepoints.

Reviewer: ```suggestion
- **Track ID**: An identifier for a unique set of objects, linking their object IDs across timepoints. Generally this describes the track of one object along the time sequence.
```

## 2. Prerequisites

### Installing `colorizer-data`

> From a command terminal, clone this repository and run the following commands to install dependencies. This will install the necessary libraries for the example scripts and the latest release of `colorizer-data`. (You may want to do this from a virtual Python environment-- see [venv](https://docs.python.org/3/library/venv.html) or [conda](https://docs.conda.io/en/latest/) for more information.)

Reviewer 2: Do you think we should have the install instructions use a virtual environment? I know in my early python days, I wasn't fluent with virtual environments and I often got myself into wild error states from conflicting dependency, and at that phase I was also likely to directly copy / paste provided commands into my terminal without following other text recommendations (which may be more of a me problem lol 😭 )

Author: I can reference some of the environment tutorial code that Philip wrote for NucMorph!

Author: Added `venv` instructions!

> **_NOTE_**: You must be on Python version 3.9 or above. The installation may fail unexpectedly on older versions of Python.

```bash
git clone https://github.com/allen-cell-animated/colorizer-data.git
cd colorizer-data/documentation/getting_started_guide

python -m pip install -r ./requirements.txt
```

## 3. Expected formats for raw data

For this tutorial, we'll be working with sample data included in the [`getting_started_guide/raw_dataset`](./getting_started_guide/raw_dataset/) directory.

This dataset is a simplified example of raw, pre-processed segmentation data. The data was generated using the [`generate_raw_data.py` script](./getting_started_guide/scripts/generate_data.py), which generates a **CSV file** with columns for object IDs, track IDs, times, centroids, features (volume/height), and paths to the segmentation images. The **segmentation images** are 2D images in the OME-TIFF format, encoding the locations of segmented objects.

> Your files may be in a different format or have 3D segmentation images, in which case it will need to be transformed. Generally, we recommend:

Reviewer: I don't want this to suggest that people have to do something special with their 3D segmentation images, since the utility code is prepared to do max projection to flatten 3d to 2d..  
```suggestion
If your segmentation images are 3D, you may choose to flatten them to 2D or use our provided utilites to do so. 
 Your tracked object data may be in a different format, in which case it will need to be transformed to work well with our utilities. Generally, we recommend:
```

| object_id | track_id | time | centroid_x | centroid_y | area | radius | location | segmentation_path |
| ----------- | ---------- | ------ | ------------ | ------------ | -------- | -------- | ------------------- | --- |
| 1 | 0 | 0 | 33 | 110 | 706.9 | 15 | middle | frame_0.tiff |
| 2 | 1 | 0 | 67 | 100 | 804.2 | 16 | middle | frame_0.tiff |
| 3 | 2 | 0 | 100 | 108 | 804.2 | 16 | middle | frame_0.tiff |
| 4 | 3 | 0 | 133 | 88 | 706.9 | 15 | middle | frame_0.tiff |
| 5 | 4 | 0 | 167 | 101 | 804.2 | 16 | middle | frame_0.tiff |
| 6 | 0 | 1 | 33 | 121 | 530.9 | 13 | bottom | frame_1.tiff |
| 7 | 1 | 1 | 67 | 113 | 804.2 | 16 | middle | frame_1.tiff |
| ... | ... | ... | ... | ... | ... | ... | ... | ... |

Each of the segmentation images is an OME-TIFF image containing the IDs of the segmented objects:

_Frame 0 of the example dataset, as viewed in FIJI. Contrast has been increased for easier viewing._

![Frame 0 of the example dataset. The background is black, labeled `value=0`. There are five circles of various diameters and positions, with IDs starting at 30 and increasing to 34.](./getting_started_guide/assets/sample-segmentation.png)

> > **_NOTE:_** Note that `value=0` is used to represent the background in the segmentation images. For simplicity, we recommend starting object IDs at `1` to avoid conflicts with the background value.

Reviewer: ```suggestion
> **_NOTE:_** A value of `0` is used to represent the background in the segmentation images. For simplicity, we recommend starting object IDs at `1` to avoid conflicts with the background value.
```

## 4. Processing data

Timelapse Feature Explorer reads data in the format specified by the [`DATA_FORMAT`](./documentation/DATA_FORMAT.md) document. We'll use the utilities provided by `colorizer-data` to convert to this format.

### Processing script

> Start an interactive Python session. Make sure you are at `./documentation/getting_started_guide/` in the cloned repository.

Reviewer: Idea: convert all this to a .ipynb file, and drop the test data in the same directory or a subdirectory.  Then the ipynb file can just load the files directly as local data without even worrying about explaining directory names and extra cd commands..

Author: Done, whew....

```bash
# If not already in the `getting_started_guide` directory:
cd /documentation/getting_started_guide/

python
```

Paste the following steps into the terminal. (Alternatively, you can also create a Python script, copy in the code below, and run it. The full script can be found as [`process_data.py` in the `scripts` directory](./getting_started_guide/scripts/process_data.py).)

#### 1. Import dependencies and load the dataset into a pandas DataFrame

```python
from bioio import BioImage
import pandas as pd

from colorizer_data.types import DATETIME_FORMAT
from colorizer_data.utils import (
    remap_segmented_image,
)
from colorizer_data.writer import (
    ColorizerDatasetWriter,
    ColorizerMetadata,
    FeatureInfo,
    FeatureType,
)

# Load the dataset
data: pd.DataFrame = pd.read_csv("raw_dataset/data.csv")
```

#### 2. Configure the writer and data columns

```python

> # Define column names

Reviewer: ```suggestion
# Define column names as they appear in the loaded csv.  Some of these are required and others are the extra calculated features for use in the colorizer.
```
I suggest adding very brief comments saying which are optional, required, and colorizable features.  Maybe good to say it elsewhere but right here is where people will be looking at the code so a good place to reinforce it.

```python
# Turn core data columns into a numpy array, to be saved by the writer.
tracks = data[TRACK_ID_COLUMN].to_numpy()
times = data[TIMES_COLUMN].to_numpy()
centroids_x = data[CENTROIDS_X_COLUMN].to_numpy()
centroids_y = data[CENTROIDS_Y_COLUMN].to_numpy()

writer.write_data(
    tracks=tracks,
    times=times,
    centroids_x=centroids_x,
    centroids_y=centroids_y,
)
```

#### 4. Write the features

Features can be one of three types:

1. **Continuous** features are used for floating-point numbers.
2. **Discrete** features are used for integers.
3. **Categorical** features are used for string-based labels. Note that there can only be a max of 12 categories for a categorical feature.

The `FeatureInfo` class is used to provide metadata about each feature, such as its label, key, type, and units if applicable.

```python
areas = data[AREA_COLUMN].to_numpy()
locations = data[LOCATION_COLUMN].to_numpy()
radii = data[RADIUS_COLUMN].to_numpy()

# Additional metadata can be provided for each feature, which will be shown
# when interacting with it in the viewer.
area_info = FeatureInfo(
    label="Area",
    key="area",
    type=FeatureType.CONTINUOUS,
    unit="px²",

> )

Reviewer: seeing this here makes me think it could be elegant to just put the numpy array in the FeatureInfo class, and then write_feature can be passed just a FeatureInfo alone!  Like `FeatureInfo(data=areas, label="Area" , ...)` something to ponder for later..

radius_info = FeatureInfo(
    label="Radius",
    key="radius",
    # Discrete features are used for integers.
    type=FeatureType.DISCRETE,
    unit="px",
)
location_info = FeatureInfo(
    label="Location",
    key="location",
    # Categorical features are used for string-based labels.
    type=FeatureType.CATEGORICAL,
    # Categories can be auto-detected from the data, or provided manually
    # if you want to preserve a specific order for the labels.
    categories=["top", "middle", "bottom"],
)
writer.write_feature(areas, area_info)
writer.write_feature(radii, radius_info)
writer.write_feature(locations, location_info)
```

#### 5. Write the images

> The `ColorizerDatasetWriter` writes images as PNGs with encoded object IDs. You can see more about what this looks like and how it works in our [data format documentation](./DATA_FORMAT.md#5-frames). As previously noted, all object IDs must be unique too, so this next section will perform three tasks:

Reviewer: ```suggestion
The `ColorizerDatasetWriter` needs to prepare images for fast visualization by converting them to PNGs with encoded object IDs. You can see more about what this looks like and how it works in our [data format documentation](./DATA_FORMAT.md#5-frames). As previously noted, all object IDs must be unique too, so this next section will perform three tasks:
```

1. Load in the image data from the segmentation images
2. Remap the object IDs to be unique across all timepoints
3. Write the remapped images using `ColorizerDatasetWriter.write_image()`

```python
# Group data by the timestamp
data_grouped_by_time = data.groupby(TIMES_COLUMN)
frame_paths = []

for frame_num, frame_data in data_grouped_by_time:

>     # Get the path to the image and load it.

Reviewer: ```suggestion
    # Get the path to the segmentation image and load it.
```

    frame_path = frame_data.iloc[0][SEGMENTED_IMAGE_COLUMN]

>     segmentation_image = BioImage("raw_dataset/" + frame_path).get_image_data(

Reviewer: Nit: maybe this should use pathlib and save "raw_dataset" as a constant so this code looks like:
```suggestion
    segmentation_image = BioImage(source_dataset_dir / frame_path).get_image_data(
```

Author: Changed, thanks!

        "YX", S=0, T=0, C=0
    )
    # NOTE: For datasets with 3D segmentations, you may need to flatten the data into
    # 2D images. Typically, it's simplest to do so with a max projection, but may vary

>     # based on your data. Replace the above line with the following:

Reviewer: 👍

    #
    # segmentation_image = bioio.BioImage(frame_path).get_image_data("ZYX", S=0, T=0, C=0)
    # segmentation_image = segmentation_image.max(axis=0)
    #
    # Remap the segmented so object IDs are unique across all timepoints.
    (remapped_segmentations, _lut) = remap_segmented_image(
        segmentation_image, frame_data, OBJECT_ID_COLUMN, INDEX_COLUMN
    )
    # Write the new segmentation image.
    frame_prefix = "frame_"
    frame_suffix = ".png"
    writer.write_image(remapped_segmentations, frame_num, frame_prefix, frame_suffix)

>     frame_paths.append(frame_prefix + str(frame_num) + frame_suffix)

Reviewer: Observation for later: users would not have to do this line if write_image returned this string value `frame_prefix + str(frame_num) + frame_suffix` (since write_image was passed in all those params)

Author: Good point, thank you

writer.set_frame_paths(frame_paths)
```

#### 6. Write the dataset and any additional metadata

```python
# Define the metadata for this dataset.
metadata = ColorizerMetadata(
    name="Example dataset",
    description="An example dataset for the Timelapse Feature Explorer.",
    author="Jane Doe et al.",
    dataset_version="v1.0",
    # The width and height of the original segmentations, in any arbitrary units.
    # This will control the scale bar in the viewer.
    frame_width=100,
    frame_height=100,
    frame_units="nm",
    # Time elapsed between each frame capture, in seconds.
    frame_duration_sec=1,
)

# Write the final dataset
writer.write_manifest(metadata=metadata)
```

> Once the steps are run, the new dataset will be found in the `processed_dataset` directory.

Reviewer: Let's add a celebratory line here like:
Congratulations! If you got this far, you have a complete dataset ready for use in the TFE viewer.

## 5. Viewing the dataset

> Now that the dataset is processed, we can view it in the Timelapse Feature Explorer!

Reviewer: 👍

> Our public release of Timelapse Feature Explorer is designed to load datasets hosted on a web server. To load **local datasets and files**, you'll need to run a **local instance** of the viewer. We'll cover steps for both options.

Reviewer: ```suggestion
Timelapse Feature Explorer is designed to load datasets using http(s), which means your data must be hosted on a web server or you can run a "local web server". To load **local datasets and files**, you'll need to run a **local instance** of the viewer. We'll cover steps for both options.
```

### Viewing datasets via the web

> **_NOTE:_** To use a dataset with our public build of Timelapse Feature Explorer, the dataset must be accessible using the HTTPS protocol (e.g., `https://example.com/your-dataset/`). If you need to use HTTP, run a local instance of the viewer.

For this tutorial, you can load a pre-processed example copy of the dataset, which we've hosted on GitHub. You can access it at this URL: `https://raw.githubusercontent.com/allen-cell-animated/colorizer-data/doc/getting-started-guide/documentation/getting_started_guide/example/processed_dataset/manifest.json`

If you have updated the dataset files or want to use a local dataset, skip to the next section ([Installing Timelapse Feature Explorer locally](#installing-timelapse-feature-explorer-locally)).

> #### Opening your dataset

Reviewer: ```suggestion
#### Opening your dataset with data hosted via https
```

> If you have updated the dataset files or want to use a local dataset, skip to the next section ([Installing Timelapse Feature Explorer locally](#installing-timelapse-feature-explorer-locally)).

Author: Should I not show users how to do this, and force them to only learn to run a local instance?

1. Open Timelapse Feature Explorer at [https://timelapse.allencell.org](https://timelapse.allencell.org).

2. Click the **Load** in the header and paste in the following URL: `https://raw.githubusercontent.com/allen-cell-animated/colorizer-data/doc/getting-started-guide/documentation/getting_started_guide/example/processed_dataset/manifest.json`

![The Load button on the Timelapse Feature Explorer header, next to the Help dropdown.](./getting_started_guide/assets/load-button.png)

> **_NOTE:_** You can either provide the URL of the directory containing a `manifest.json` or the full URL path of a `.json` file that follows the [manifest specification](./DATA_FORMAT.md#dataset). We recommend specifying the full URL path that includes the `manifest.json`.

Click **Load** in the popup menu to load the dataset. The viewer should appear with the dataset loaded!

> ### Installing Timelapse Feature Explorer locally

Reviewer: ```suggestion
### Opening your dataset from a local directory over http
```

To view our **locally converted dataset**, we'll also need to run a **local version** of the Timelapse Feature Explorer.

<details>
<summary><b>[Why do we need to run a local instance of the viewer?]</b></summary>

---
The public version of TFE is served over HTTPS, which is a secure protocol. For security reasons, HTTPS pages cannot load HTTP content, which means that the public version of TFE can only access web content hosted on HTTPS.

We can run a local server to serve our local files over HTTP, but we need to also run TFE on HTTP to access them.

---

</details>

#### 1. TFE installation

Install Node (v20 or above) from [nodejs.org](https://nodejs.org/). Once Node is installed, **open a command terminal** in the directory you want to install TFE into. Run the following commands to clone the repository and install the dependencies:

```bash
git clone https://github.com/allen-cell-animated/timelapse-colorizer.git
cd timelapse-colorizer
npm install
npm run start
```

> You can now run `npm run start` at anytime to start the viewer. By default, it will be mounted at `http://localhost:5173`.

Reviewer: ```suggestion
You can now run `npm run start` at anytime in this directory to start the viewer. By default, it will be mounted at `http://localhost:5173`.
```
